src/librustc_trans/trans/intrinsic.rs

```diff
@@ -989,3 +1033,162 @@ fn with_overflow_intrinsic<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
+    let tcx = bcx.tcx();
+    let arg_tys = match callee_ty.sty {
+        ty::TyBareFn(_, ref f) => {
+            bcx.tcx().erase_late_bound_regions(&f.sig.inputs())
+        }
+        _ => unreachable!()
+    };
+
+    enum Cmp { Eq, Ne, Lt, Le, Gt, Ge }
+    let comparison = match name {
+        "simd_eq" => Some(Cmp::Eq),
+        "simd_ne" => Some(Cmp::Ne),
+        "simd_lt" => Some(Cmp::Lt),
+        "simd_le" => Some(Cmp::Le),
+        "simd_gt" => Some(Cmp::Gt),
+        "simd_ge" => Some(Cmp::Ge),
+        _ => None
+    };
+
+    macro_rules! require {
+        ($cond: expr, $($fmt: tt)*) => {
+            if !$cond {
+                bcx.sess().span_err(call_info.span, &format!($($fmt)*));
+                return C_null(llret_ty)
+            }
+        }
+    }
+
+    if let Some(cmp) = comparison {
+        assert_eq!(arg_tys.len(), 2);
+        // we need nominal equality here, not LLVM (structural)
+        // equality
+        require!(arg_tys[0] == arg_tys[1],
+                 "SIMD comparison intrinsic monomorphised with different input types");
+        require!(arg_tys[0].is_simd(tcx),
+                 "SIMD comparison intrinsic monomorphised for non-SIMD argument type");
+        require!(ret_ty.is_simd(tcx),
+                 "SIMD comparison intrinsic monomorphised for non-SIMD return type");
+
+        let in_len = arg_tys[0].simd_size(tcx);
+        let out_len = ret_ty.simd_size(tcx);
+        require!(in_len == out_len,
+                 "SIMD comparison intrinsic monomorphised for non-SIMD argument type");
+        require!(llret_ty.element_type().kind() == llvm::Integer,
+                 "SIMD comparison intrinsic monomorphised with non-integer return");
+
+        let llarg_ty0 = val_ty(llargs[0]);
+        let llin_elem_ty = llarg_ty0.element_type();
+        return if llin_elem_ty.kind() == llvm::Integer {
+            let is_signed = arg_tys[0].simd_machine_type(tcx).unwrap().is_signed();
+            let op = match cmp {
+                Cmp::Eq => llvm::IntEQ,
+                Cmp::Ne => llvm::IntNE,
+                Cmp::Lt => if is_signed {llvm::IntSLT} else {llvm::IntULT},
+                Cmp::Le => if is_signed {llvm::IntSLE} else {llvm::IntULE},
+                Cmp::Gt => if is_signed {llvm::IntSGT} else {llvm::IntUGT},
+                Cmp::Ge => if is_signed {llvm::IntSGE} else {llvm::IntUGE},
+            };
+            SExt(bcx, ICmp(bcx, op, llargs[0], llargs[1], call_debug_location), llret_ty)
+        } else {
+            let op = match cmp {
+                Cmp::Eq => llvm::RealOEQ,
+                Cmp::Ne => llvm::RealUNE,
+                Cmp::Lt => llvm::RealOLT,
+                Cmp::Le => llvm::RealOLE,
+                Cmp::Gt => llvm::RealOGT,
+                Cmp::Ge => llvm::RealOGE,
+            };
+            SExt(bcx, FCmp(bcx, op, llargs[0], llargs[1], call_debug_location), llret_ty)
+        };
```

Comment:
Could this leverage [`compare_scalar_types`](https://github.com/rust-lang/rust/blob/ed49bad0ccb0e9ee7e5ebea69d72a98bed08f77f/src/librustc_trans/trans/base.rs#L314) instead?